presto-main/src/main/java/io/prestosql/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -521,8 +521,9 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
 
             if (node.getType() == INNER) {
                 canonicalCriteria.stream()
-                        .filter(clause -> types.get(clause.getLeft()).equals(types.get(clause.getRight())))
```

Comment:
This should be safe to remove (equi should be on same type), so you can extract this to a preparatory cleanup